tests/app/notifications/test_process_notification.py

```diff
@@ -249,7 +253,42 @@ def test_persist_notification_sets_daily_limit_cache_if_one_does_not_exists(
             reference="ref2",
         )
 
-        mock_set.assert_called_once_with(str(service.id) + "-2016-01-01-count", 1, ex=86400)
+        assert mocker.call(str(service.id) + "-2016-01-01-count", 1, ex=86400) in mock_set.call_args_list
+
+
+@pytest.mark.parametrize(
+    "notification_type, recipient",
+    (
+        (LETTER_TYPE, "123 Fake Street, FA11KE"),
+        (SMS_TYPE, "447700900986"),
+        (EMAIL_TYPE, "fake@notify.works"),
+    ),
+)
+@freeze_time("2016-01-01 11:09:00.061258")
+def test_persist_notification_sets_rate_limit_per_notification_type(
+    mocker, notify_api, notify_db_session, notification_type, recipient
+):
+    service = create_service(restricted=False)
+    template = create_template(service=service)
+    api_key = create_api_key(service=service)
+    mocker.patch("app.notifications.process_notifications.redis_store.get", return_value=None)
+    mock_set = mocker.patch("app.notifications.process_notifications.redis_store.set")
+    with set_config(notify_api, "REDIS_ENABLED", True):
+        persist_notification(
+            template_id=template.id,
+            template_version=template.version,
+            recipient=recipient,
+            service=template.service,
+            personalisation={},
+            notification_type=notification_type,
+            api_key_id=api_key.id,
+            key_type=api_key.key_type,
+            reference="ref2",
+        )
+
+    assert (
+        mocker.call(str(service.id) + f"-{notification_type}-2016-01-01-count", 1, ex=86400) in mock_set.call_args_list
```

Comment:
can we make this just one f string rather than a weird concatenation of a f string and a normal string 🙏

Reply:
also we've now got three tests that kinda just interact weirdly.

for trial and live, make sure incr is called for the generic count
for trial and live, make sure set is called for the generic count
for the three notification types, make sure set is called for that specific count

im not entirely sure what would make these three tests cleaner, but i wonder if the notification type check can be parametrised in to the existing two tests? that might be overkill

also, i'd like to assert on the entire call args list, to ensure we're not setting additional incorrect keys anywhere